### DISCUSSION TITLE
Implementing UWG pre-processors as QGIS processing tools

### DISCUSSION THREAD

**Discussion by aleksandraradecka1** [UMEP-dev/UMEP#367] [Python]
During the last seminar Fredrik mentioned the possibility of implementing the remaining UWG pre-processors as QGIS processing tools (only UWG Reclassifier, Land cover Reclassifier and UWG Prepare are left). Do you know approximately when will it happen?

The tools would be very helpful for making more comprehensive, automated tests of the model's performance and the influence of each element on the final result.

I would be very greatful for the help.

**Comment by biglimp** [comment]
The UWG Reclassifier and Land Cover Reclassifier is not suitable for the processor since it requires first an upload a vector polygon file and then a reclassification made by the user. That is not possible in a processing tool so this has to be left in the menu-based UMEP. However, the UWG Prepare tool could be migrated with relatively little effort. Hopefully I can make this happen soon (if no one else wants to try...).

**Reply by aleksandraradecka1** [reply]
Thank you for the answer. 

I understand. Regarding both Reclassifiers - I came up with the idea of using Field Calculator instead as the work is only about creating one column based on the existing one. 

For UWG Prepare - I think you maybe the best person to do this as the tool links many of the existing ones. I however volunteer to test it quickly.

**Comment by biglimp** [comment]
I found some time and added UWG Prepare in UMEP for Processing. It is still experimental and only available via [the development release](https://umep-docs.readthedocs.io/en/latest/Getting_Started.html#installing-development-release-could-be-unstable). Please try it out and see if it works.

**Reply by aleksandraradecka1** [reply]
Thank you very much ! I will test it tomorrow (max. Thursday) and let you know.

**Comment by aleksandraradecka1** [comment]
Dear Fredrik, 

I managed to install the development release of UMEP for Processing and found UWG Prepare tool in it. 

However when I run the tool I obtained an error (screenshot attached). I think the last letter of the path to the Vector Polygon Grid is missing somewhere in the code. I got the same error for different file names and extensions. I also used the indicated Vector Polygon Grid in other tool (Land Cover Fraction (Grid)) and it was working just fine.  

Cheers, 
Aleksandra

<img width="660" alt="UWG_prepare_error" src="https://user-images.githubusercontent.com/44131961/153214730-e46216a3-89cd-4a3c-b664-adda6224af3a.png">

**Reply by biglimp** [reply]
Ahh, I noticed the same issue when having very long filenames. Thy to reduce the actual name of the shapefile and see if it works.

**Comment by aleksandraradecka1** [comment]
I tried with a new shorter filename and also shorter path in general. It now looks like this: "C:/umep/pgrid.shp" and the error is unfortunately the same.

**Reply by biglimp** [reply]
Could you send a screendump so that I see the exact error.

**Comment by aleksandraradecka1** [comment]
Sure, it looks like this:
<img width="665" alt="UWG_prepare_error2" src="https://user-images.githubusercontent.com/44131961/153234845-656c34a0-b5fb-497f-885d-d4f193c77906.png">

**Reply by aleksandraradecka1** [reply]
If there is anything I can further test or help you please let me know.

**Comment by biglimp** [comment]
Thanks. I need to figure out why and when we are loosing a p in shp in certain cases. The error in not consistent as I can run the model if I shorten the filename. I will get back to you.

**Reply by aleksandraradecka1** [reply]
Of course, good luck !

**Reply by biglimp** [reply]
I just submitted a new version to the repository. Try now.

**Reply by aleksandraradecka1** [reply]
I did. It now gives some other type of error: 
<img width="664" alt="zero_error" src="https://user-images.githubusercontent.com/44131961/153383121-89dd373b-71dd-4f83-b054-04933c00abe8.png">

**Reply by biglimp** [reply]
Two questions: 

1. Do you get an error when you run the UWG Prepare plugin in the menu-based UMEP using the same data?
2. Do you have any grid where you have no urban typology polygon present?

**Reply by aleksandraradecka1** [reply]
OK. It seems we have a turning point:

1. When I run the tool without buildings layer it works well both in menu-based UMEP and UMEP for processing.

2. When I run the tool with buildings layer in menu-based UMEP I obtain the error:
<img width="1044" alt="error_buildings" src="https://user-images.githubusercontent.com/44131961/153401119-1ee8584d-54b5-420e-aeb7-c8bcbf8d54e7.png">
I checked this tempdata folder and indeed there is an SHP file called "intersected" which is being created while I run the UWG Prepare tool. The layer seems to be an intersection of buliding and polygon grid.

**Reply by biglimp** [reply]
Thanks. Could you attach your input data (grid, typology layer, morhology text file and land cover text file) so that I can see if I can find what is wrong.

**Reply by aleksandraradecka1** [reply]
Of course. Here they are:
[input_data.zip](https://github.com/UMEP-dev/UMEP/files/8040779/input_data.zip)

**Reply by biglimp** [reply]
I just submitted a new version which hopefully will take care of your error above. 
With that said I noticed two things with your data that is problematic:
1. Your urban typology has no info (NULL) in the i **UWGTime** attribute. This will break UWG later on. I will fix the **UWGReclassifier** so that it is not possible to skip that when you reclassify your data. It should be either _Pre80_, _Pst80_ or _New_.
2. Your **btype** layer is not what is expected that should be used as typology layer. The typology layer should be of local scale and not for individual buildings. See the attached screen dump for my test data. Here buildings with similar typopogies are mapped together in a neighboorhood scale. I will also add this info later on when I work on the documentation and tutorial for UWG.

![image](https://user-images.githubusercontent.com/11908356/153439623-bd6efa7a-5703-4512-b3b1-84d4b4e610a8.png)

**Reply by aleksandraradecka1** [reply]
Dear Fredrik, 

first of all - thank you very much for the work done last two days ! I installed the new version of UMEP for processing, filled UWGTime attribute and now the tool is working. 

I will allow myself to inquire a bit more into the data part of using the tool:

1. I understand the concept of neighboorhoods you described above. I am now just wondering how did you obtain this kind of data for your project. Did you create them yourself based on buildings layer? Or maybe you prepared the layer in an automated way throught buffers etc.? Or maybe this kind of layer is ready available in your country? 

2. Climate zones in the tool are American what is of course not a surprise. Can you maybe recommend some good literature source/webpage for translating them to European climate zones?

3. Both in Image Morphometric Calculator and Land Cover Fractions (Grid) the result includes isotropic and anisotropic data. However later, in UWG Prepare, only the first mentioned are used. Can anisotropic be somehow incorporated into the model as well? Sorry if the question is too basic.

**Reply by biglimp** [reply]
Great! Thanks for all the testing.
1. This particular dataset is derived from swedish national data sources and then manipulated further manually. I could recommend looking into urban atlas (https://land.copernicus.eu/local/urban-atlas) or open street map (www.osm.org) as a base for your typology layer. Maybe you also have some national data source that can be used e.g. separating between commercial, residential and industry in cities.
2. I have not looked into this yet but maybe a translation to Köppen climate zones would be appropriate in the future.
3. Not a basic question. With some manual work, the anisotrophic files could be used based on e.g. wind directions. Then I would recommend to aggregate a number of wind directions in the anisotrophic files e.g. if wind is coming from 270 degrees, I would aggregate 255 to 285 and create an new "fake" isotrophic from these aggragations. Hope I made my self understandable here.

**Reply by aleksandraradecka1** [reply]
Thanks!

1. Yes, I downloaded Urban Atlas but I decided it is too coarse. Now I think it may serve as one alternative I should test. And yes, I also checked OSM, but the problem with it is that the landuse layer (at least on my researach area) does not encompass all buildings, and also the OSM's landuse does not directly inform about building types. I think I will do the same as you did - transform the national buildings layer. 

3. I think I understand: I may use "standard" isotropic file if I do not know where the wind comes from on my researach area. If I however know and it is e.g. 270 degrees, the aggregation of 255 and 285 will just give more precise results?

**Reply by biglimp** [reply]
2. Yes, using only one wind direction e.g. 270 can produce strange results.

**Reply by aleksandraradecka1** [reply]
OK. I will do some tests and pay attention to this aspect. 

Thank you once again !

**Comment by aleksandraradecka1** [comment]
Dear Fredrik, 

I had a break from UMEP and I started my work today again. Unfortunately, I encountered a bunch of errors while trying to use the UWG processor. 

These are:
- green rectangle - occurs when there are no buildings in a grid element (I of course changed the single buildings layer to building typologies and filled the time attribute)
- orange rectangle - occurs when the calculations are done correctly, but this requires that I manually add "_UWG" to the .epw file name for a given grid element
- blue rectangle - occurs when I do not change the file name
- purple rectangle - hard for me to say what is the reason

Could you please advise me what to do? 

Aleksandra
<img width="668" alt="errors_uwg" src="https://user-images.githubusercontent.com/44131961/157057564-f7a37e3a-d474-4f64-8ea7-1ed0cb569dcb.png">
<img width="427" alt="data_uwg" src="https://user-images.githubusercontent.com/44131961/157057567-e56636ed-8a50-4bf7-b0f8-c2ab92ca8470.png">

**Reply by biglimp** [reply]
Let us take one error at the time.
Green Rectangle:
open the Python console and just run one grid where this happens. Then we can see what error message that pops out from UWG in detail.

Also make sure that you are using the latest development release.

**Reply by aleksandraradecka1** [reply]
I reinstalled the UMEP for Processing plugin to the newest development release and it behaves the same. I enclose the screenshot of my Python console and input files.
[data_grid_el1.zip](https://github.com/UMEP-dev/UMEP/files/8198846/data_grid_el1.zip)
<img width="1006" alt="green_rectangle_error" src="https://user-images.githubusercontent.com/44131961/157065389-578b0e7f-60a6-440e-8c6d-6f82e0c4a112.png">

**Reply by biglimp** [reply]
It seems like this model cannot run if you dont have any buildings within the grid. Test to set the bldDensity to something very small, e.g. 0.001. If this work I will change the preproccessing code accordingly.

**Reply by aleksandraradecka1** [reply]
I changed the value as you recommended and obtained the info that the three parameters must be less than 1 (screenshot attached), so I further changed the file by typing "treeCover,0.999" and the error "transformed" to the type indicated by blue rectangle (screenshot attached). Therefore we can state the the solution works for the green problem. 
<img width="401" alt="bld_1" src="https://user-images.githubusercontent.com/44131961/157068404-5a01eccb-d0bd-48aa-811e-7b20d6b2391b.png">
<img width="397" alt="bld_2" src="https://user-images.githubusercontent.com/44131961/157068415-e104e8c4-1b18-4e1e-80ca-91f1c29b05d4.png">

**Reply by biglimp** [reply]
Great. Now run grid 7 (or grid 1 with new fractions) with the python console open and see what the error message show.

**Reply by aleksandraradecka1** [reply]
Running grid element 1 results in the creation of two .epw files neither of which having a name expected by the algorithm - "prep_1" and "rep_1_UWG" while the algorithm expects "prep_1_UWG".
<img width="1050" alt="grid_el_1_error" src="https://user-images.githubusercontent.com/44131961/157208521-2635be9c-4998-46f0-b267-115a717ee549.png">

**Reply by biglimp** [reply]
Do you tick in to save UMEP format?

Could you attach grid 1 and your epw input file and tell about your settings so that I can test.

**Reply by aleksandraradecka1** [reply]
Previously I did not tick. I now checked running the tool with this options and results are quite the same (Python console screenshot attached).
[grid_1.zip](https://github.com/UMEP-dev/UMEP/files/8205010/grid_1.zip)
<img width="750" alt="error_ticked" src="https://user-images.githubusercontent.com/44131961/157218198-c118c29b-07d8-4f97-870a-f5330db1c5e5.png">

**Reply by biglimp** [reply]
I also need grid_el_1.uwg to test

**Reply by aleksandraradecka1** [reply]
Of course:
[grid_1_corrected.zip](https://github.com/UMEP-dev/UMEP/files/8207674/grid_1_corrected.zip)

**Reply by biglimp** [reply]
Ojojoj. This is a very strange error occurring! 

Happily, it seams not to be on our end, but on the Ladybug UWG people side of things. 

When you use the letter **p** in the beginning of your filename for your uwg-file the output from UWG is loosing that **p**, so in your case you should have a file called **rep_1_UWG.ewp** in the same folder as where you have your uwg-file. If you change your uwg-filename to e.g. **test** and try again it should work for you. It did in my case.

I recommend that you submit an issue to the ladybug github about this (https://github.com/ladybug-tools/uwg).

Please also check that you can run the model if you change filename.

**Reply by aleksandraradecka1** [reply]
Yes, it is true. A filename starting with "t" works well, and of course I will submit the issue to ladybug in a minute. Thank you

I would just like to ask you if the file "metdata_UMEP.txt" was produced by UWG Processor in your case? In mine it is produced for selected runs and I do not see the pattern when. This file as you probably remember is needed for UWG Analyzer.

**Reply by biglimp** [reply]
It should be created when you tick-in that you want to save **UMEP-specific format...** which is needed to use the UWG Analyser. Let me know if the file is not created.

Your last error (purple) could maybe be solved if you decrease simulation time to e.g. 150 or 100.

**Reply by aleksandraradecka1** [reply]
I tried and it is indeed created when ticked-in, thank you. I will also try with different times. 

I have just submitted an issue to ladybug: https://github.com/ladybug-tools/uwg/issues/268 I hope I explained everything clearly.

**Reply by PetitMote** [reply]
I can confirm the problem with the zero building grid: you need at least 0.001 bldDensity to make it work.
